### PR TITLE
[Draft] Add VM metrics

### PR DESCRIPTION
**Pull request by mikker**
Closes elastic/apm-agent-ruby#387

### CODE REVIEW

File: lib/elastic_apm/metrics/vm.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+      def sample
+        @sampler.sample
+      end
+
+      def collect
+        return unless sampler
+
+        sample.each_with_object({}) do |(key, value), snap|
+          snap[:"#{SCOPE}#{key}"] = value
```
Comment (axw) [inline lib/elastic_apm/metrics/vm.rb]: I don't think we need to be capturing _all_ of the GC statistics, and we should be a bit more explicit about the names we choose to report. I suggest these names as a start:

 - ruby.gc.count
 - ruby.gc.time
 - ruby.thread.count (or ruby.threads? Java is using `.count`, but I'm not sure if that's the ideal convention)

Do you intend to add heap usage too?

File: lib/elastic_apm/metrics/vm.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    # @api private
+    class VM
+      def initialize(_config)
+        @total_time = 0
+      end
+
+      # rubocop:disable Metrics/MethodLength
+      def collect
+        stat = GC.stat
+        @total_time += GC::Profiler.total_time
```
Comment (axw) [inline lib/elastic_apm/metrics/vm.rb]: I think we should only report `ruby.gc.time` if the GC profiler is enabled.

File: lib/elastic_apm/metrics/vm.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        GC::Profiler.clear
+        thread_count = Thread.list.count
+
+        {
+          'ruby.gc.count': stat[:count],
+          'ruby.gc.time': @total_time,
+          'ruby.heap.live': stat[:heap_live_slots],
```
Comment (axw) [inline lib/elastic_apm/metrics/vm.rb]: Repeating my message from Slack, for posterity: let's call this `ruby.heap.slots.live`, and `ruby.heap.slots.free` below.

File: lib/elastic_apm/metrics/vm.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+          'ruby.heap.free': stat[:heap_free_slots],
+          'ruby.allocations.total': stat[:total_allocated_objects],
```
Comment (axw) [inline lib/elastic_apm/metrics/vm.rb]: this should probably have a ".heap" in it right? how about `ruby.heap.objects.allocated`?

**Comment by mikker** [comment]
I'll include the docs and changelog entry in this PR. Just wanted to get the impl. right first.

**Comment by mikker** [comment]
Thanks once again for helping out, @axw!

**Comment by mikker** [comment]
@bmorelli25 Docs update spotted!

File: docs/metrics.asciidoc [hunk 5]
```diff
@@ -12,11 +12,15 @@ You can adjust the interval by setting <<config-metrics-interval,`metrics_interv
 
 The metrics will be stored in the `apm-*` index and have the `processor.event` property set to `metric`.
 
+[float]
+[[metrics-system]]
+=== System metrics
+
 **Note:** Metrics from the Ruby agent are Linux only for now.
```
Comment (bmorelli25) [inline docs/metrics.asciidoc]:
```suggestion
NOTE: Metrics from the Ruby agent are Linux only for now.
```

Let's turn this into a real "note"

**Comment by mikker** [comment]
Continued in #487 because Jenkins

**Pull request closed without merging**